protocols/bittorrent.ts

```diff
@@ -0,0 +1,94 @@
+import { Static } from '@sinclair/typebox'
+import Protocol, { Ctx, SyncOptions } from './interfaces'
+import { BitTorrentProtocolFields } from '../api/schemas'
+
+import path from 'node:path'
+import { cp } from 'node:fs/promises'
+
+import { TorrentManager } from 'bt-fetch'
+
+export interface BitTorrentProtocolOptions {
+  path: string
+}
+
+export class BitTorrentProtocol implements Protocol<Static<typeof BitTorrentProtocolFields>> {
+  options: BitTorrentProtocolOptions
+  manager: TorrentManager | null
+
+  constructor (options: BitTorrentProtocolOptions) {
+    this.options = options
+    this.manager = null
+  }
+
+  async load (): Promise<void> {
+    const folder = this.options.path
+    this.manager = new TorrentManager({ folder })
+  }
+
+  async unload (): Promise<void> {
+    await this.manager?.destroy()
+  }
+
+  async sync (id: string, folderPath: string, options?: SyncOptions, ctx?: Ctx): Promise<Static<typeof BitTorrentProtocolFields>> {
+    ctx?.logger.info('[bittorrent] Sync Start')
+
+    const link = `bittorent://${id}/`
+
+    if (this.manager === null) {
+      throw new Error('[bittorrent] Torrent Manager Not Initialized')
+    }
+
+    const manager: TorrentManager = this.manager
+
+    // Create keypair from the site ID and the seed key
```

Comment:
does this generate the same `publicKey` given the same ID? in my head, this is what's happening as that's how the stop seeding in L56 makes sense

Reply:
The way it works is the TorrentManager will generate a seed key in the storage if one doesn't exist. The `name` will then be used along side the seed to derive a keypair for the site. This process is deterministic. I make use of the [derive-key module](https://www.npmjs.com/package/derive-key)

You can see the code for deriving here: https://github.com/RangerMauve/bt-fetch/blob/default/torrent-manager.js#L570